File: problems/0025-reverse-nodes-in-k-group/python/solution.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def build(vals):
    dummy = ListNode()
    cur = dummy
    for v in vals:
        cur.next = ListNode(v)
        cur = cur.next
    return dummy.next


def to_list(head):
    out = []
    while head:
        out.append(head.val)
        head = head.next
    return out


def reverseKGroup(head, k):
    """Reverse Nodes in k-Group — https://leetcode.com/problems/reverse-nodes-in-k-group/"""
    node = head
    count = 0
    while node and count < k:
        node = node.next
        count += 1
    if count < k:
        return head
    prev = reverseKGroup(node, k)  # process the remainder first
    cur = head
    for _ in range(k):
        nxt = cur.next
        cur.next = prev
        prev = cur
        cur = nxt
    return prev


def reverseKGroupArr(vals, k):
    return to_list(reverseKGroup(build(vals), k))

File: problems/0025-reverse-nodes-in-k-group/go/solution.go
package solution

type ListNode struct {
	Val  int
	Next *ListNode
}

func build(vals []int) *ListNode {
	dummy := &ListNode{}
	cur := dummy
	for _, v := range vals {
		cur.Next = &ListNode{Val: v}
		cur = cur.Next
	}
	return dummy.Next
}

func toSlice(head *ListNode) []int {
	out := []int{}
	for n := head; n != nil; n = n.Next {
		out = append(out, n.Val)
	}
	return out
}

// Reverse Nodes in k-Group — https://leetcode.com/problems/reverse-nodes-in-k-group/
func reverseKGroup(head *ListNode, k int) *ListNode {
	node := head
	count := 0
	for node != nil && count < k {
		node = node.Next
		count++
	}
	if count < k {
		return head
	}
	prev := reverseKGroup(node, k) // process the remainder first
	cur := head
	for i := 0; i < k; i++ {
		nxt := cur.Next
		cur.Next = prev
		prev = cur
		cur = nxt
	}
	return prev
}

func reverseKGroupArr(vals []int, k int) []int {
	return toSlice(reverseKGroup(build(vals), k))
}
